app/scripts/keyring-controller.js

```diff
@@ -0,0 +1,566 @@
+    const wallet = configManager.getWallet() // old style vault
+    const vault = configManager.getVault() // new style vault
+
+    return {
+      seedWords: this.configManager.getSeedWords(),
+      isInitialized: (!!wallet || !!vault),
+      isUnlocked: !!this.key,
+      isDisclaimerConfirmed: this.configManager.getConfirmedDisclaimer(), // AUDIT this.configManager.getConfirmedDisclaimer(),
+      unconfTxs: this.configManager.unconfirmedTxs(),
+      transactions: this.configManager.getTxList(),
+      unconfMsgs: messageManager.unconfirmedMsgs(),
+      messages: messageManager.getMsgList(),
+      selectedAddress: address,
+      selectedAccount: address,
+      shapeShiftTxList: this.configManager.getShapeShiftTxList(),
+      currentFiat: this.configManager.getCurrentFiat(),
+      conversionRate: this.configManager.getConversionRate(),
+      conversionDate: this.configManager.getConversionDate(),
+      keyringTypes: this.keyringTypes.map((krt) => krt.type()),
+      identities: this.identities,
+    }
+  }
+
+  setStore (ethStore) {
+    this.ethStore = ethStore
+  }
+
+  createNewVaultAndKeychain (password, entropy, cb) {
+    this.createNewVault(password, entropy, (err) => {
+      if (err) return cb(err)
+      this.createFirstKeyTree(password, cb)
+    })
+  }
+
+  createNewVaultAndRestore (password, seed, cb) {
+    if (typeof password !== 'string') {
+      return cb('Password must be text.')
+    }
+
+    if (!bip39.validateMnemonic(seed)) {
+      return cb('Seed phrase is invalid.')
+    }
+
+    this.clearKeyrings()
+
+    this.createNewVault(password, '', (err) => {
+      if (err) return cb(err)
+      this.addNewKeyring('HD Key Tree', {
+        mnemonic: seed,
+        numberOfAccounts: 1,
+      }, (err) => {
+        if (err) return cb(err)
+        const firstKeyring = this.keyrings[0]
+        const accounts = firstKeyring.getAccounts()
+        const firstAccount = accounts[0]
+        const hexAccount = normalize(firstAccount)
+        this.configManager.setSelectedAccount(hexAccount)
+        this.setupAccounts(accounts)
+
+        this.emit('update')
+        cb()
+      })
+    })
+  }
+
+  migrateAndGetKey (password) {
+    let key
+    const shouldMigrate = !!this.configManager.getWallet() && !this.configManager.getVault()
+
+    return this.loadKey(password)
+    .then((derivedKey) => {
+      key = derivedKey
+      this.key = key
+      return this.idStoreMigrator.oldSeedForPassword(password)
+    })
+    .then((serialized) => {
+      if (serialized && shouldMigrate) {
+        const keyring = this.restoreKeyring(serialized)
+        this.keyrings.push(keyring)
+        this.configManager.setSelectedAccount(keyring.getAccounts()[0])
+      }
+      return key
+    })
+  }
+
+  createNewVault (password, entropy, cb) {
+    const configManager = this.configManager
+    const salt = this.encryptor.generateSalt()
+    configManager.setSalt(salt)
+
+    return this.migrateAndGetKey(password)
```

Comment:
this should just be `this.loadKey(password)` instead since we never need to care about an old vault when creating a new vault.